File: Python/easy/83. Remove Duplicates from Sorted List.py
"""
Given a sorted linked list, delete all duplicates such that each element appear only once.
Example 1:
Input: 1->1->2
Output: 1->2
Example 2:
Input: 1->1->2->3->3
Output: 1->2->3
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def deleteDuplicates(self, head: 'ListNode') -> 'ListNode':
        if not head or not head.next:
            return head
        valSet = set()
        valSet.add(head.val)
        curNode = head.next
        prevNode = head
        while curNode != None:
            if curNode.val not in valSet:
                valSet.add(curNode.val)
                curNode = curNode.next
                prevNode = prevNode.next
            else:
                curNode = curNode.next
                prevNode.next = curNode
        return head

File: Java/easy/83. Remove Duplicates from Sorted List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode curNode = head;
        while (curNode != null) {
            while (curNode.next != null && curNode.next.val == curNode.val) curNode.next = curNode.next.next;
            curNode = curNode.next;
        }
        return head;
    }
}